build.sbt

```diff
@@ -78,70 +127,130 @@ lazy val common = createProject("reactive-lib-common", "common")
     crossScalaVersions := Vector(Versions.scala211, Versions.scala212)
   )
 
+lazy val serviceDiscoveryAssemblySettings = Vector(
+  libraryDependencies ++= Seq(
+    "com.typesafe.akka" %% "akka-actor"          % Versions.akka              % "provided",
+    "com.typesafe"       % "config"              % Versions.typesafeConfig    % "provided",
+    "com.lightbend"     %% "service-locator-dns" % Versions.serviceLocatorDns
+  ),
+  assemblyExcludeLocal("common"),
+  assemblyShadeRules in assembly ++= Seq(
+    ShadeRule.rename("akka.io.AsyncDnsResolver**" -> "com.lightbend.rp.internal.akka.io.AsyncDnsResolver").inAll,
+    ShadeRule.rename("ru.**" -> "com.lightbend.rp.internal.ru.@0").inAll,
+    ShadeRule.rename("spray.**" -> "com.lightbend.rp.internal.spray.@0").inAll,
+    ShadeRule.rename("com.typesafe.**" -> "com.lightbend.rp.internal.com.typesafe.@0").inAll,
+    ShadeRule.rename("com.lightbend.dns.locator.**" -> "com.lightbend.rp.internal.dns.locator.@0").inAll,
+    ShadeRule.rename("org.**" -> "com.lightbend.rp.internal.org.@0").inAll,
+    ShadeRule.rename("scala.concurrent.java8.**" -> "com.lightbend.rp.internal.scala.concurrent.java8.@0").inAll,
+    ShadeRule.rename("scala.compat.java8.**" -> "com.lightbend.rp.internal.scala.compat.java8.@0").inAll
```

Comment:
I'm shading the `service-locator-dns` and `akka-dns` libs because `akka-dns` isn't published to Sonatype. This way, we can still provide a good experience without having to add a resolver to the user's project.

Reply:
I think I'd like to sit down with you to go through the shaded names since I have these questions:

- what is the `org.**` patterns used for? The match seems to be very broad.
- why do we need to shade `scala.concurrent.java8.**` and `scala.compat.java8.**`?
- I think we should not shade `service-locator-dns` and `akka-dns` to allow end users to override the versions of these two if they choose to. Not shading also means we can independently release these two independently of `reactive-lib`.

For now perhaps the SBT Reactive App can add the resolvers, and we just document in `reactive-lib` the need to add the resolvers if used directly without SBT Reactive App.

In the long run, I'd suggest approaching `akka-dns` owner about additional publishing to Sonatype, and perhaps we can do PR for him if necessary.

Reply:
Was the [test output](https://github.com/lightbend/reactive-lib/pull/5#issuecomment-344088190) you posted done using the patterns above?

I think the nesting of `io/netty/io/netty` or `org/org/asynchttpclient/` is incorrect:

```
com/lightbend/rp/internal/io/netty/io/netty/util/
com/lightbend/rp/internal/io/netty/io/netty/util/collection/
com/lightbend/rp/internal/io/netty/io/netty/util/concurrent/

com/lightbend/rp/internal/org/org/asynchttpclient/
com/lightbend/rp/internal/org/org/asynchttpclient/channel/
```

Reply:
As discussed, we're going to shade `akka-dns` and `service-locator-dns` for now, with the intention of using upstream dependencies and/or a module system in the future.